pkg/storage/cache_test.go

```diff
@@ -171,9 +167,145 @@ func TestCheckCacheKeyContextualTuplesWithConditionsOrdering(t *testing.T) {
 	})
 	require.NoError(t, err)
 
+	require.NotEqual(t, key1, key2)
+}
+
+func TestCheckCacheKeyConsidersConditionContext(t *testing.T) {
+	storeID := ulid.Make().String()
+	modelID := ulid.Make().String()
+
+	struct1, err := structpb.NewStruct(map[string]interface{}{
+		"key1": "foo",
+		"key2": "bar",
+	})
+	require.NoError(t, err)
+
+	struct2, err := structpb.NewStruct(map[string]interface{}{
+		"key1": "foo",
+		"key3": "baz",
+	})
+	require.NoError(t, err)
+
+	jonContextOne := tuple.NewTupleKeyWithCondition(
+		"document:2",
+		"admin",
+		"user:jon",
+		"some_condition",
+		struct1,
+	)
+	jonContextTwo := tuple.NewTupleKeyWithCondition(
+		"document:2",
+		"admin",
+		"user:jon",
+		"some_condition",
+		struct2,
+	)
+
+	tupleKey := tuple.NewTupleKey("document:x", "viewer", "user:jon")
+
+	key1, err := GetCheckCacheKey(&CheckCacheKeyParams{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tupleKey,
+		ContextualTuples:     []*openfgav1.TupleKey{jonContextOne},
+	})
+	require.NoError(t, err)
+
+	key2, err := GetCheckCacheKey(&CheckCacheKeyParams{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tupleKey,
+		ContextualTuples:     []*openfgav1.TupleKey{jonContextTwo},
+	})
+	require.NoError(t, err)
+
+	require.NotEqual(t, key1, key2)
+}
+
+func TestCheckCacheKeyConditionContextOrderAgnostic(t *testing.T) {
+	storeID := ulid.Make().String()
+	modelID := ulid.Make().String()
+
+	struct1, err := structpb.NewStruct(map[string]interface{}{
+		"key1": "foo",
+		"key2": "bar",
+	})
+	require.NoError(t, err)
+
+	// same object, keys in different order
+	struct2, err := structpb.NewStruct(map[string]interface{}{
+		"key2": "bar",
+		"key1": "foo",
+	})
+	require.NoError(t, err)
+
+	jonContextOne := tuple.NewTupleKeyWithCondition(
+		"document:2",
+		"admin",
+		"user:jon",
+		"some_condition",
+		struct1,
+	)
+	jonContextTwo := tuple.NewTupleKeyWithCondition(
+		"document:2",
+		"admin",
+		"user:jon",
+		"some_condition",
+		struct2,
+	)
+
+	tupleKey := tuple.NewTupleKey("document:x", "viewer", "user:jon")
+
+	key1, err := GetCheckCacheKey(&CheckCacheKeyParams{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tupleKey,
+		ContextualTuples:     []*openfgav1.TupleKey{jonContextOne},
+	})
+	require.NoError(t, err)
+
+	key2, err := GetCheckCacheKey(&CheckCacheKeyParams{
+		StoreID:              storeID,
+		AuthorizationModelID: modelID,
+		TupleKey:             tupleKey,
+		ContextualTuples:     []*openfgav1.TupleKey{jonContextTwo},
+	})
+	require.NoError(t, err)
+
 	require.Equal(t, key1, key2)
 }
 
+func TestCheckCacheKeyContextualTuplesConditionOrderMatters(t *testing.T) {
```

Comment:
I don't understand this test, and this test name :/

We _are_ sorting the contextual tuples (line 40 of `hasher.go`), so the order in which customers send them should not matter.

I also don't understand why this test is passing. I expect line 306 to be `require.Equal`.

Reply:
This is about to change. You're right that this should be `require.Equal`, that was a bug in the new logic I added in this PR which added `condition` as part of the cache key _without_ factoring it in as part of the sorting logic. The contextual tuples were only sorted based on user/object/relation. I'm about to push some changes that will also factor the `condition` into the sorting, and this test will be fixed.

There's also the case where multiple contextual_tuples have the _same_ condition name but different contexts, which may also need to be added into the sort logic; I haven't fully worked that one out yet.